File: tests/test_simd_f64.js
/* test_simd_f64.js — dyna:simd double-precision (f64) kernels over
 * Float64Array. Run: dynajs (CONFIG_NATIVE_MODULES=y) tests/test_simd_f64.js
 *
 * JS Number IS f64, so the JS reference below computes the SAME IEEE-754 ops as
 * the kernels:
 *   - f64Scale / f64Axpy are BIT-EXACT vs JS (===). axpy is a non-fused
 *     multiply-then-add on every ISA, matching JS `y[i] + a*x[i]` (JS has no
 *     fused multiply-add), so === holds.
 *   - f64Max / f64Min are exact (===).
 *   - f64Sum / f64Dot REORDER additions, so they are checked with a
 *     reorder-aware (condition-number) tolerance: |got-ref| <= 1e-12 * Σ|terms|.
 */
import { f64Sum, f64Dot, f64Max, f64Min, f64Scale, f64Axpy } from "dyna:simd";

let n = 0;
function assert(c, m) { n++; if (!c) throw new Error("assertion failed: " + m); }

/* sizes hit every f64 vector width (NEON/SSE2=2, AVX=4, AVX-512=8) plus
 * multi-vector bodies and odd tails. */
const SIZES = [0, 1, 2, 3, 4, 5, 7, 8, 9, 15, 16, 17, 100, 1000];
const rnd64 = (len, seed) => Float64Array.from({ length: len }, (_, i) => ((i * 7 + seed) % 17) - 8.5);

/* ---------- f64Sum / f64Dot: reorder-aware tolerance ---------- */
for (const len of SIZES) {
    const a = rnd64(len, 1), b = rnd64(len, 2);
    let s = 0, sCond = 0, d = 0, dCond = 0;
    for (let i = 0; i < len; i++) {
        s += a[i]; sCond += Math.abs(a[i]);
        d += a[i] * b[i]; dCond += Math.abs(a[i] * b[i]);
    }
    assert(Math.abs(f64Sum(a) - s) <= 1e-12 * Math.max(1, sCond), "f64Sum len=" + len);
    assert(Math.abs(f64Dot(a, b) - d) <= 1e-12 * Math.max(1, dCond), "f64Dot len=" + len);
}
assert(f64Sum(new Float64Array(0)) === 0, "f64Sum empty = 0");
assert(f64Dot(new Float64Array(0), new Float64Array(0)) === 0, "f64Dot empty = 0");

/* ---------- f64Max / f64Min: exact; empty throws ---------- */
for (const len of SIZES) {
    const a = rnd64(len, 3);
    if (len === 0) {
        for (const [name, fn] of [["f64Max", f64Max], ["f64Min", f64Min]]) {
            let t = false; try { fn(a); } catch { t = true; }
            assert(t, name + ": empty array throws");
        }
        continue;
    }
    let mx = a[0], mn = a[0];
    for (let i = 1; i < len; i++) { if (a[i] > mx) mx = a[i]; if (a[i] < mn) mn = a[i]; }
    assert(f64Max(a) === mx, "f64Max len=" + len);
    assert(f64Min(a) === mn, "f64Min len=" + len);
}

/* ---------- f64Scale: in-place, bit-exact, returns the same array ---------- */
for (const len of SIZES) {
    for (const s of [2.5, -3, 0.1, 0]) {
        const a = rnd64(len, 4), c = Float64Array.from(a);
        const r = f64Scale(c, s);
        assert(r === c, "f64Scale returns its array, len=" + len);
        for (let i = 0; i < len; i++) assert(c[i] === a[i] * s, "f64Scale[" + i + "] len=" + len + " s=" + s);
    }
}

/* ---------- f64Axpy: in-place non-fused, bit-exact, returns y ---------- */
for (const len of SIZES) {
    for (const al of [2.5, -3, 0.1]) {
        const y0 = rnd64(len, 5), x = rnd64(len, 6), y = Float64Array.from(y0);
        const r = f64Axpy(y, al, x);
        assert(r === y, "f64Axpy returns y, len=" + len);
        for (let i = 0; i < len; i++) assert(y[i] === y0[i] + al * x[i], "f64Axpy[" + i + "] len=" + len + " a=" + al);
    }
}

/* ---------- type + shape errors ---------- */
{
    let t = false; try { f64Dot(new Float64Array(3), new Float64Array(4)); } catch { t = true; }
    assert(t, "f64Dot length mismatch throws");
    t = false; try { f64Axpy(new Float64Array(3), 1, new Float64Array(4)); } catch { t = true; }
    assert(t, "f64Axpy length mismatch throws");
    t = false; try { f64Sum(new Float32Array(4)); } catch { t = true; } /* bpe 4 != 8 */
    assert(t, "f64Sum on a Float32Array throws");
    t = false; try { f64Sum(new Uint8Array(8)); } catch { t = true; }
    assert(t, "f64Sum on a Uint8Array throws");
    t = false; try { f64Sum([1, 2, 3]); } catch { t = true; }
    assert(t, "f64Sum on a plain array throws");
}

/* ---------- reentrancy: a scalar arg's valueOf runs BEFORE the buffer is
 * resolved (coerce-first discipline), so it can't corrupt the operation. ---- */
{
    const a = new Float64Array([1, 2, 3, 4]);
    let calls = 0;
    f64Scale(a, { valueOf() { calls++; return 2; } });
    assert(calls === 1 && a[0] === 2 && a[3] === 8, "f64Scale coerces scalar first");

    const y = new Float64Array([1, 2, 3, 4]), x = new Float64Array([10, 20, 30, 40]);
    let c2 = 0;
    f64Axpy(y, { valueOf() { c2++; return 3; } }, x);
    assert(c2 === 1 && y[0] === 31 && y[3] === 124, "f64Axpy coerces scalar first");
}

print("test_simd_f64: all tests passed (" + n + " assertions)");

File: tests/test_simd_f64.c
/*
 * test_simd_f64.c -- regression + differential test for the shared SIMD
 * double-precision (f64) array kernels across every ISA in the dispatch table:
 *   f64_sum, f64_dot, f64_min, f64_max, f64_scale, f64_axpy.
 *
 * Guards the two historical SIMD-kernel traps in this repo:
 *   1. HEAP-OOB READ: a reduction that seeds a full SIMD-width vector
 *      (vld1q_f64/_mm_loadu_pd/_mm256_loadu_pd/_mm512_loadu_pd/svld1) BEFORE
 *      checking n reads past the buffer for 0 < n < width. Every f64 reduction
 *      here early-returns a scalar loop for small n; this test proves it.
 *   2. CORRECTNESS across ISAs and small/boundary/tail sizes.
 *
 * Method: place exactly n doubles flush against an unmapped guard page, so any
 * read past x[n) SIGSEGVs (works WITHOUT ASan -- important under qemu). Compare
 * every kernel to a sequential scalar reference for n = 1..129.
 *   - f64_sum / f64_dot REORDER additions vs a sequential loop, so they are NOT
 *     bit-identical: checked with a RELATIVE-ERROR tolerance (1e-12).
 *   - f64_min / f64_max / f64_scale / f64_axpy ARE exact: checked bit-for-bit.
 *     (axpy is a non-fused mul-then-add on every ISA, matching the scalar ref.)
 *
 * Build+run natively (arm64 NEON / x86 host). -ffp-contract=off keeps the C
 * reference non-fused so the exact axpy check is valid on every compiler:
 *   clang -O2 -ffp-contract=off -Isrc tests/test_simd_f64.c \
 *     src/dyna-simd-core.c src/dyna-simd-scalar.c src/dyna-simd-neon.c \
 *     src/dyna-simd-sse42.c src/dyna-simd-avx2.c src/dyna-simd-avx512.c \
 *     src/dyna-simd-sve.c -lpthread -lm -o /tmp/tsf64 && /tmp/tsf64
 * Exercise the x86 kernels under emulation (build into /tmp, never host .obj):
 *   docker run --rm --platform linux/amd64 -e QEMU_CPU=Haswell -v "$PWD":/src \
 *     -w /src dynascript-dev:amd64 bash -c '<build line above>'
 *   (QEMU_CPU: Nehalem => SSE2, Haswell => AVX, Skylake-Server => AVX-512.)
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <sys/mman.h>
#include <unistd.h>
#include "dyna-simd-kernels.h"

static double ref_sum(const double *x, size_t n) {
    double a = 0.0;
    for (size_t i = 0; i < n; i++) a += x[i];
    return a;
}
static double ref_dot(const double *a, const double *b, size_t n) {
    double s = 0.0;
    for (size_t i = 0; i < n; i++) s += a[i] * b[i];
    return s;
}
static double ref_max(const double *x, size_t n) {
    double m = x[0];
    for (size_t i = 1; i < n; i++) if (x[i] > m) m = x[i];
    return m;
}
static double ref_min(const double *x, size_t n) {
    double m = x[0];
    for (size_t i = 1; i < n; i++) if (x[i] < m) m = x[i];
    return m;
}

/* strict bit-for-bit equality (no NaN in the test data, so this is exactness) */
static int d_bits_eq(double a, double b) { return memcmp(&a, &b, sizeof a) == 0; }

/* reordered-reduction tolerance: |got-ref| / max(1,|ref|) <= 1e-12 */
static int rel_ok(double got, double ref) {
    double denom = fabs(ref) < 1.0 ? 1.0 : fabs(ref);
    return fabs(got - ref) / denom <= 1e-12;
}

/* deterministic pseudo-random double in ~[-153, 153] with a fractional part */
static double gen(size_t i, int trial, unsigned salt) {
    unsigned h = (unsigned)i * 2654435761u + (unsigned)trial * 40503u + salt * 2246822519u;
    return (double)((int)(h % 4000u) - 2000) / 13.0;
}

int main(void) {
    simd_init();
    unsigned long long c = cpu_features();
    printf("caps=0x%llx SSE42=%d AVX2=%d AVX512=%d NEON=%d SVE=%d\n", c,
           !!(c & CPU_SSE42), !!(c & CPU_AVX2), !!(c & CPU_AVX512F),
           !!(c & CPU_NEON), !!(c & CPU_SVE));

    long pg = sysconf(_SC_PAGESIZE);
    /* two guard regions: some kernels touch two arrays (a/b, x/out, y/x). */
    char *g0 = mmap(0, 2 * pg, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    char *g1 = mmap(0, 2 * pg, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (g0 == MAP_FAILED || g1 == MAP_FAILED) { perror("mmap"); return 2; }
    if (mprotect(g0 + pg, pg, PROT_NONE) || mprotect(g1 + pg, pg, PROT_NONE)) {
        perror("mprotect"); return 2;
    }

    int fails = 0;
    double yorig[130], expect[130];

    for (size_t n = 1; n <= 129; n++) {
        for (int trial = 0; trial < 24; trial++) {
            /* arrays flush against the guard page: any read past [n) faults */
            double *A = (double *)(g0 + pg - n * sizeof(double));
            double *B = (double *)(g1 + pg - n * sizeof(double));
            double s = ((trial % 7) - 3) * 0.5 + 1.25; /* scalar for scale/axpy */

            /* ---- f64_sum (relative) ---- */
            for (size_t i = 0; i < n; i++) A[i] = gen(i, trial, 1);
            if (!rel_ok(simd.f64_sum(A, n), ref_sum(A, n))) { printf("f64_sum n=%zu WRONG\n", n); fails++; }

            /* ---- f64_max / f64_min (exact) ---- */
            if (!d_bits_eq(simd.f64_max(A, n), ref_max(A, n))) { printf("f64_max n=%zu WRONG\n", n); fails++; }
            if (!d_bits_eq(simd.f64_min(A, n), ref_min(A, n))) { printf("f64_min n=%zu WRONG\n", n); fails++; }

            /* ---- f64_dot (relative), a=A flush g0, b=B flush g1 ---- */
            for (size_t i = 0; i < n; i++) B[i] = gen(i, trial, 2);
            if (!rel_ok(simd.f64_dot(A, B, n), ref_dot(A, B, n))) { printf("f64_dot n=%zu WRONG\n", n); fails++; }

            /* ---- f64_scale, separate out (x=A flush g0, out=B flush g1) exact ---- */
            for (size_t i = 0; i < n; i++) A[i] = gen(i, trial, 3);
            simd.f64_scale(B, A, s, n);
            for (size_t i = 0; i < n; i++)
                if (!d_bits_eq(B[i], A[i] * s)) { printf("f64_scale(out) n=%zu i=%zu WRONG\n", n, i); fails++; break; }
            /* ---- f64_scale, in-place (out == x) exact ---- */
            for (size_t i = 0; i < n; i++) expect[i] = A[i] * s;
            simd.f64_scale(A, A, s, n);
            for (size_t i = 0; i < n; i++)
                if (!d_bits_eq(A[i], expect[i])) { printf("f64_scale(inplace) n=%zu i=%zu WRONG\n", n, i); fails++; break; }

            /* ---- f64_axpy: y=A flush g0, x=B flush g1, exact non-fused ---- */
            for (size_t i = 0; i < n; i++) { A[i] = gen(i, trial, 4); B[i] = gen(i, trial, 5); yorig[i] = A[i]; }
            simd.f64_axpy(A, s, B, n);
            for (size_t i = 0; i < n; i++) {
                double p = s * B[i];           /* non-fused, matches the kernel */
                double want = yorig[i] + p;
                if (!d_bits_eq(A[i], want)) { printf("f64_axpy n=%zu i=%zu WRONG\n", n, i); fails++; break; }
            }

            if (fails > 20) { printf("... too many failures, stopping\n"); goto done; }
        }
    }
done:
    if (fails == 0)
        printf("OK: f64 sum/dot/min/max/scale/axpy correct + no guard-page fault (no OOB), n=1..129\n");
    else
        printf("FAILS=%d\n", fails);
    return fails ? 1 : 0;
}
